File: CS39000-CP2/week5/Mravi.cpp
#include <bits/stdc++.h>
#include <iostream>
#include <math.h>
#include <stdio.h>
#include <tuple>
#include <utility>

using namespace std;

double DFS(int v, int p, vector<vector<int [3]>> adj, int needs[], int t) {
    int amount = 0;
    // code for handling leaf nodes goes here, return
    if (needs[v] != -1) {
        return t = 1 ? sqrt(needs[v]) : needs[v];
    }

    for (auto &edge : adj.at(v)) {  
        if (edge[0] == p) {
            continue;
        }

        amount += DFS(edge[0], v, adj, needs, edge[2]);
    }

    return amount;
}

int main() {
    int N;
    scanf("%d", &N);
    
    int needs[N];
    vector<vector<int [3]>> adj;
    for (int i = 0; i < N; i++) {
        vector<int[3]> v;
        adj.push_back(v);
    }

    for (int i = 0; i < N - 1; i++) {
        int A, B, X, T;
        scanf("%d %d %d %d", &A, &B, &X, &T);
    
        int a1[3] = {B, X, T};
        adj.at(A - 1).push_back(a1);
        int a2[3] = {A, X, T};
        adj.at(B - 1).push_back(a2);
    }
    for (int i = 0; i < N; i++) {
        cin >> needs[i];
    }

    printf("%f", DFS(0, 0, adj, needs, 0));

    return 0;
}

File: CS39000-CP2/week5/Mravi.java
import java.util.*;
import java.io.*;
import java.lang.*;

public class Mravi {
  public static double amount = 0;

  public static double DFS(int v, double x, int t, int p, ArrayList<ArrayList<int[]>> adj, int[] needs) {
    double total = 0;

    if (needs[v] != -1) {
      total = t == 1 ? Math.sqrt((double) needs[v]) : needs[v];
      return total / x;
    }

    double subtree = 0;
    for (int[] edge : adj.get(v)) {  
      if (edge[0] == p) {
          continue;
      }

      total = Math.max(total, DFS(edge[0], ((double) edge[1]) / 100, edge[2], v, adj, needs));
    }

    if (t == 1) {
      total = Math.sqrt(total);
    }
    return total / x;
  }

  public static void main(String[] args) throws IOException {
    Scanner scan = new Scanner(System.in);

    int N = scan.nextInt();

    int[] needs = new int[N];
    ArrayList<ArrayList<int[]>> adj = new ArrayList<ArrayList<int[]>>(N);
    for (int i = 0; i < N; i++) {
      adj.add(new ArrayList<int[]>());
    }

    for (int i = 0; i < N - 1; i++) {
      int A = scan.nextInt() - 1;
      int B = scan.nextInt() - 1;
      int X = scan.nextInt();
      int T = scan.nextInt();

      adj.get(A).add(new int[]{B, X, T});
      adj.get(B).add(new int[]{A, X, T});
    }
    for (int i = 0; i < N; i++) {
      needs[i] = scan.nextInt();
    }

    System.out.println(DFS(0, 1.0, 0, 0, adj, needs));
  }
}